Decompiler: Add support for INT_REM in jumptable recovery.

LukeSerne opened the pull request:
This commit adds `INT_REM` handling to the `JumpBasic` class, which allows switch statements depending on expressions like `x % 10` to be properly recovered. Previously, a warning was emitted about there being "too many branches". This slightly improves the third problem mentioned in https://github.com/NationalSecurityAgency/ghidra/issues/2628#issuecomment-1869907358. 

**Screenshots**
These screenshots show the `main` function of the `switch` example from `GhidraDocs/GhidraClass/Advanced/Examples` after auto-analysis.

Before applying the patch
![image](https://github.com/NationalSecurityAgency/ghidra/assets/32435769/14313f59-50e5-406b-b164-e0d6b11d3e99)

After applying the patch
![image](https://github.com/NationalSecurityAgency/ghidra/assets/32435769/c7ea9a8e-ac99-4acb-a6cd-ea150c5ae24c)

LukeSerne committed:
Decompiler: Add support for INT_REM in jumptable recovery.
This commit adds INT_REM handling to JumpBasic, which allows switch statements
depending on expressions like "x % 10" to be properly recovered. Previously, a
warning was emitted about there being "too many branches".

mumbel commented on Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:
```diff
@@ -510,7 +510,7 @@ uintb JumpBasic::backup2Switch(Funcdata *fd,uintb output,Varnode *outvn,Varnode
 }
 
-/// If the Varnode has a restricted range due to masking via INT_AND, the maximum value of this range is returned.
+/// If the Varnode has a restricted range due to masking via INT_AND or by INT_REM with a constnant, the maximum value of this range is returned.
```
`constant`

LukeSerne replied:
Thanks for spotting the typo! It's fixed now.

Nemoumbra commented:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/62447396/db4048a1-2912-4b56-a84f-a6c41e90b679)

This won't help with my case... *sniff*

Nemoumbra commented:
@LukeSerne I have no clue which program I was studying back then.😭
And I have not tried it with latest Ghidra either, maybe it'd be fine there.

LukeSerne committed:
Decompiler: Better max jumptable index calculation
This commit adds a fallback to use the non-zero mask to determine the
maximum value a varnode can take. It also slightly improves the case
where a varnode is defined by a MULTIEQUAL by recursively calling the
getMaxValue function.

jobermayr commented on Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:
```diff
@@ -528,24 +532,28 @@ uintb JumpBasic::getMaxValue(Varnode *vn)
-    if (i == op->numInput()) {
-      maxValue = coveringmask( maxValue );
-      maxValue = (maxValue + 1) & calc_mask(vn->getSize());
+      uintb submax = getMaxValue(subvn);
```
[crash.xml](https://github.com/user-attachments/files/28980062/crash.xml)

I get a segmentation fault after an infinite loop (only three Varnodes are processed; see eg. frames 3-5, 6-8, 9-11, ...):
```
/tmp/ghidra/build/dist/ghidra_12.2_DEV/Ghidra/Features/Decompiler/os/linux_x86_64/decomp_dbg
[decomp]> restore crash.xml
crash.xml successfully loaded: Intel/AMD 64-bit x86
[decomp]> load function cupslog_printf
Function cupslog_printf: 0x00400cb2
[decomp]> decompile
Decompiling cupslog_printf
Speicherzugriffsfehler     (Speicherabzug geschrieben) /tmp/ghidra/build/dist/ghidra_12.2_DEV/Ghidra/Features/Decompiler/os/linux_x86_64/decomp_dbg
```

```
Program received signal SIGSEGV, Segmentation fault.
0x0000000000435ace in std::vector<ghidra::Varnode*, std::allocator<ghidra::Varnode*> >::size (this=<error reading variable: Cannot access memory at address 0x7fffff7fefe8>) at /usr/include/c++/15/bits/stl_vector.h:1117
1117          size() const _GLIBCXX_NOEXCEPT
Missing separate debuginfos, use: zypper install glibc-debuginfo-2.43-3.1.x86_64 binutils-debuginfo-2.45-4.1.x86_64 libsframe2-debuginfo-2.45-4.1.x86_64 libz1-debuginfo-1.3.1-2.2.x86_64 libzstd1-debuginfo-1.5.7-4.2.x86_64 libstdc++6-debuginfo-16.1.1+git8886-1.1.x86_64 libgcc_s1-debuginfo-16.1.1+git8886-1.1.x86_64
(gdb) bt
#0  0x0000000000435ace in std::vector<ghidra::Varnode*, std::allocator<ghidra::Varnode*> >::size (this=<error reading variable: Cannot access memory at address 0x7fffff7fefe8>)
    at /usr/include/c++/15/bits/stl_vector.h:1117
#1  0x0000000000435b74 in std::vector<ghidra::Varnode*, std::allocator<ghidra::Varnode*> >::operator[] (this=0x2f507f0, __n=0) at /usr/include/c++/15/bits/stl_vector.h:1263
#2  0x0000000000434b43 in ghidra::PcodeOp::getIn (this=0x2f507a0, slot=0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/include/ghidra/op.hh:157
#3  0x00000000005eafad in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:538
#4  0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#5  0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#6  0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#7  0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#8  0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#9  0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#10 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#11 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#12 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#13 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#14 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#15 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#16 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#17 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#18 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#19 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#20 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#21 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#22 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#23 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#24 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#25 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#26 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#27 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#28 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#29 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#30 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#31 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#32 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#33 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#34 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#35 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#36 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#37 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#38 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#39 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#40 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#41 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#42 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#43 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#44 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#45 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#46 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#47 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#48 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#49 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#50 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#51 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#52 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#53 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f50360) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#54 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f508a0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
#55 0x00000000005eafbd in ghidra::JumpBasic::getMaxValue (vn=0x2f4f9f0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/jumptable.cc:539
```

LukeSerne replied:
Thanks for the report! I see how the implementation could lead to infinite recursion. I just pushed what I think is a fix.

I couldn't test it properly with your debug xml though. First I got some errors about non-existent context variables `ctxAX`, `ctxAH` and `ctxAL`, and when I commented out those lines in the xml, I got the below error when trying to load the function.

```
[decomp]> rest /tmp/crash.xml
/tmp/crash.xml successfully loaded: Intel/AMD 64-bit x86
[decomp]> lo fu cupslog_printf
Low-level ERROR: Bad instruction count exceeded:
    ...
    Instruction at (ram,0x00400956) overlaps instruction at (ram,0x00400955)

Unable to proceed with function: cupslog_printf
```

Could you test it and confirm that the issue no longer exists?

jobermayr replied:
> I couldn't test it properly with your debug xml though. First I got some errors about non-existent context variables `ctxAX`, `ctxAH` and `ctxAL`, and when I commented out those lines in the xml, I got the below error when trying to load the function.

Sorry. It's from ghidra-staging with patch to fix #2266. I didn't know it affects xml file.

> 
> ```
> [decomp]> rest /tmp/crash.xml
> /tmp/crash.xml successfully loaded: Intel/AMD 64-bit x86
> [decomp]> lo fu cupslog_printf
> Low-level ERROR: Bad instruction count exceeded:
>     ...
>     Instruction at (ram,0x00400956) overlaps instruction at (ram,0x00400955)
> 
> Unable to proceed with function: cupslog_printf
> ```

Culprit for this is commit d424b6f.

> 
> Could you test it and confirm that the issue no longer exists?

With d424b6f reverted it loads and decompiles but has some
`/* WARNING: Bad instruction - Truncating control flow here */`
Maybe those WARNINGs now lead to Low-level ERRORs?

LukeSerne replied:
Hmm, interesting. Good to hear it's no longer crashing, but oof that function looks bad when decompiled. It seems like the decompiler gets very confused about the control flow, it loses track of the stack pointer and more.

Could you share the binary that contains this function? I'd be curious if the jump table recovery is the cause of the problems, and it's easier (for me) to debug that with the complete binary than with a debug xml.

jobermayr replied:
> Could you share the binary that contains this function?

[brps2brfax](https://github.com/user-attachments/files/29174149/brps2brfax.txt)

LukeSerne replied:
Thanks! The `cupslog_printf` function is quite weird indeed. It seems like `AL` is another parameter to the function, which is `0` in every call. However, the function is exported, so other programs might use it differently. If you add a register assume for the `AL` register at the start of the function (so `AL` is 0), the function does decompile quite nicely (even without this PR applied):

```c
void cupslog_printf(char *fmt, ...)

{
  char cVar1;
  undefined8 in_RCX;
  long lVar2;
  undefined8 in_RDX;
  undefined8 in_RSI;
  char *pcVar3;
  undefined8 in_R8;
  undefined8 in_R9;
  undefined4 uStack_d8;
  undefined4 uStack_d4;
  undefined1 *puStack_d0;
  undefined1 *puStack_c8;
  int iStack_bc;
  undefined1 auStack_b8 [8];
  undefined8 local_b0;
  undefined8 local_a8;
  undefined8 local_a0;
  undefined8 local_98;
  undefined8 local_90;
  
  uStack_d8 = 8;
  uStack_d4 = 0x30;
  puStack_d0 = &stack0x00000008;
  puStack_c8 = auStack_b8;
  lVar2 = -1;
  pcVar3 = &stderr_buf;
  do
  {
    if (lVar2 == 0) break;
    lVar2--;
    cVar1 = *pcVar3;
    pcVar3 += 1;
  } while (cVar1 != '\0');
  iStack_bc = ~(uint)lVar2 - 1;
  local_b0 = in_RSI;
  local_a8 = in_RDX;
  local_a0 = in_RCX;
  local_98 = in_R8;
  local_90 = in_R9;
  vsnprintf(&stderr_buf + iStack_bc, 0x1fff - (long)iStack_bc, fmt, &uStack_d8);
  return;
}
```

I'm not sure what would make a compiler emit this code. From the strings in the binary, we can see it was compiled with GCC 4.0.3 / 3.4.6, which are from 2006...

It seems like the code assumes `AL < 8`, otherwise the jump jumps into misaligned instructions. While this is not necessarily a problem on x86, it is very unusual. Those extra instructions that can be executed with `AL` all happen to be exactly 4 bytes long, and each stores one of the `XMM` registers to the stack (at this point, `RAX` points to a location on the stack).

```
               .text:00400d16 ff e2            JMP                 RDX
               .text:00400d18 0f 29 78 f1      MOVAPS              xmmword ptr [RAX + -0xf], XMM7
               .text:00400d1c 0f 29 70 e1      MOVAPS              xmmword ptr [RAX + -0x1f], XMM6
               .text:00400d20 0f 29 68 d1      MOVAPS              xmmword ptr [RAX + -0x2f], XMM5
               .text:00400d24 0f 29 60 c1      MOVAPS              xmmword ptr [RAX + -0x3f], XMM4
               .text:00400d28 0f 29 58 b1      MOVAPS              xmmword ptr [RAX + -0x4f], XMM3
               .text:00400d2c 0f 29 50 a1      MOVAPS              xmmword ptr [RAX + -0x5f], XMM2
               .text:00400d30 0f 29 48 91      MOVAPS              xmmword ptr [RAX + -0x6f], XMM1
               .text:00400d34 0f 29 40 81      MOVAPS              xmmword ptr [RAX + -0x7f], XMM0
                           LAB_00400d38                                    XREF[1]:     cupslog_printf:00400cf9(*)  
               .text:00400d38 48 89 bd 28      MOV                 qword ptr [RBP + -0xd8], RDI
                              ff ff ff
```

I think this code is very unusual. Thanks for sharing this interesting example with me! I think the problem has been resolved, so I'll resolve this conversation.

LukeSerne committed:
Decompiler: Prevent infinite recursion
By keeping track of how deep in the recursive chain we are, we can avoid
infinite recursion.
